api/src/main/java/org/jboss/el/cache/FactoryFinderCache.java

```diff
@@ -82,30 +84,20 @@ public static String loadImplementationClassName(final String factoryId, final C
             }
         }
 
-        String serviceId = "META-INF/services/" + factoryId;
+        ExpressionFactory result = null;
         // try to find services in CLASSPATH
         try {
-            InputStream is = null;
-            if (classLoader == null) {
-                is = ClassLoader.getSystemResourceAsStream(serviceId);
-            } else {
-                is = classLoader.getResourceAsStream(serviceId);
+            ServiceLoader<ExpressionFactory> serviceLoader = ServiceLoader.load(ExpressionFactory.class, classLoader);
+            Iterator<ExpressionFactory> iter = serviceLoader.iterator();
+            while (result == null && iter.hasNext()) {
+                result = iter.next();
             }
 
-            if (is != null) {
-                BufferedReader rd =
-                        new BufferedReader(new InputStreamReader(is, "UTF-8"));
-
-                String factoryClassName = rd.readLine();
-                rd.close();
-
-                if (factoryClassName != null &&
-                        !"".equals(factoryClassName)) {
-                    if (classCache != null) {
-                        classCache.put(new CacheKey(classLoader, factoryId), factoryClassName);
-                    }
-                    return factoryClassName;
+            if (result != null) {
+                if (classCache != null) {
+                    classCache.put(new CacheKey(classLoader, factoryId), result.getClass().getName());
```

Comment:
Why not keep the current classCache.put(new CacheKey(classLoader, factoryId), factoryClassName)?

Reply:
It looks like the next call won't find the entry in cache since the `factoryId` wasn't used when putting the result in the cache:
```
if (classCache != null) {
            final String value = classCache.get(new CacheKey(classLoader, factoryId));
            if (value != null) {
                if (value.equals("")) {
                    return null;
                }
                return value;
            }
        }
```

Reply:
factoryClassName was the string for the class name read in from the BufferedReader before.  Using the ServiceLoader approach, result now represents what's read in instead so result.getClass().getName() provides that read in class name now instead of the prior BufferedReader populated factoryClassName.

And the factoryId is still actually used in the put so should be ok for the cache get?

But looking back, I see I got a little to focused on just addressing it for a ExpressionFactory that I was reproducing the error with.  I imagine we'd want to keep this more general to work with any factor class look up if needed, so I adjusted this PR to account for that.